cdk/dcm_floodlight/transform.js

```diff
@@ -0,0 +1,171 @@
+/* eslint-disable no-param-reassign */
+const get = require("get-value");
+const {
+  getValueFromMessage,
+  removeUndefinedAndNullValues,
+  isDefinedAndNotNull
+} = require("rudder-transformer-cdk/build/utils");
+const { CustomError, getIntegrationsObj, isEmpty } = require("../../v0/util");
+
+// append properties to endpoint
+// eg: ${endpoint}key1=value1;key2=value2;....
+function appendProperties(endpoint, payload) {
+  Object.keys(payload).forEach(key => {
+    endpoint += `${key}=${payload[key]};`;
+  });
+
+  return endpoint;
+}
+
+// transform webapp dynamicForm custom floodlight variable
+// into {u1: value, u2: value, ...}
+function transformCustomVariable(customFloodlightVariable) {
+  const customVariable = {};
+  customFloodlightVariable.forEach(item => {
+    if (!isEmpty(item.from) && !isEmpty(item.to)) {
+      // remove u if already there
+      customVariable[
+        `u${item.from.trim().replace(/u/g, "")}`
+      ] = `[${item.to.trim()}]`;
+    }
+  });
+
+  return customVariable;
+}
+
+// valid flag should be provided [1|true] or [0|false]
+function isValidFlag(key, value) {
+  if (["true", "1"].includes(value.toString())) {
+    return 1;
+  }
+  if (["false", "0"].includes(value.toString())) {
+    return 0;
+  }
+
+  throw new CustomError(
+    `[DCM Floodlight]:: ${key}: valid parameters are [1|true] or [0|false]`,
+    400
+  );
+}
+
+function trackPostMapper(input, mappedPayload, rudderContext) {
+  const { message, destination } = input;
+  const { advertiserId, conversionEvents } = destination.Config;
+  let { activityTag, groupTag } = destination.Config;
+  let customFloodlightVariable;
+  let salesTag;
+
+  const baseEndpoint = "https://ad.doubleclick.net/ddm/activity/";
+
+  let event = getValueFromMessage(message, "event");
+  if (!event) {
+    throw new CustomError(
+      "[DCM Floodlight]:: event is required for track call",
+      400
+    );
+  }
+
+  const userAgent = get(message, "context.userAgent");
+  if (!userAgent) {
+    throw new CustomError(
+      "[DCM Floodlight]:: userAgent is required for track call",
+      400
+    );
+  }
+  rudderContext.userAgent = userAgent;
+
+  // find conversion event
+  // some() stops execution if at least one condition is passed and returns bool
+  event = event.trim().toLowerCase();
+  const conversionEventFound = conversionEvents.some(conversionEvent => {
+    if (conversionEvent.eventName.trim().toLowerCase() === event) {
```

Comment:
```suggestion
    if (get(conversionEvent, "eventName") === event.trim().toLowerCase() ) {
```
We are already securing event above with defined check

Reply:
We do not need `get` for 1st level children and we should keep the previous logic as it trims and lowercases the receiving events from the web app to bring it in common ground.

Reply:
We should consider the scenario `conversionEvent` being undefined its very well possible as we are using dynamic form

Reply:
If we check webapp code for dynamicForm it can never be undefined

Reply:
Same point as below